src/lib/operators/insert.cpp

```diff
@@ -107,136 +93,132 @@ std::shared_ptr<const Table> Insert::_on_execute(std::shared_ptr<TransactionCont
+   */
   {
-    auto scoped_lock = _target_table->acquire_append_mutex();
+    const auto append_lock = _target_table->acquire_append_mutex();
+
+    auto remaining_rows = input_table_left()->row_count();
 
     if (_target_table->chunk_count() == 0) {
       _target_table->append_mutable_chunk();
     }
 
-    start_chunk_id = _target_table->chunk_count() - 1;
-    end_chunk_id = start_chunk_id + 1;
-    auto last_chunk = _target_table->get_chunk(start_chunk_id);
-    start_index = last_chunk->size();
+    while (remaining_rows > 0) {
+      auto target_chunk_id = ChunkID{_target_table->chunk_count() - 1};
+      auto target_chunk = _target_table->get_chunk(target_chunk_id);
 
-    // If last chunk is compressed, add a new uncompressed chunk
-    if (!last_chunk->is_mutable()) {
-      _target_table->append_mutable_chunk();
-      end_chunk_id++;
-    }
+      // If the last Chunk of the target Table is either immutable or full, append a new mutable Chunk
+      if (!target_chunk->is_mutable() || target_chunk->size() == _target_table->max_chunk_size()) {
+        _target_table->append_mutable_chunk();
+        ++target_chunk_id;
+        target_chunk = _target_table->get_chunk(target_chunk_id);
+      }
 
-    auto remaining_rows = total_rows_to_insert;
-    while (remaining_rows > 0) {
-      auto current_chunk = _target_table->get_chunk(static_cast<ChunkID>(_target_table->chunk_count() - 1));
-      auto rows_to_insert_this_loop = std::min(_target_table->max_chunk_size() - current_chunk->size(), remaining_rows);
+      const auto target_chunk_num_inserted_rows =
+          std::min<size_t>(_target_table->max_chunk_size() - target_chunk->size(), remaining_rows);
 
-      // Resize MVCC vectors.
-      current_chunk->get_scoped_mvcc_data_lock()->grow_by(rows_to_insert_this_loop, MvccData::MAX_COMMIT_ID);
+      _target_chunk_ranges.emplace_back(
+          ChunkRange{target_chunk_id, target_chunk->size(),
+                     static_cast<ChunkOffset>(target_chunk->size() + target_chunk_num_inserted_rows)});
 
-      // Resize current chunk to full size.
-      auto old_size = current_chunk->size();
-      for (ColumnID column_id{0}; column_id < current_chunk->column_count(); ++column_id) {
-        typed_segment_processors[column_id]->resize_vector(current_chunk->get_segment(column_id),
-                                                           old_size + rows_to_insert_this_loop);
-      }
+      // Grow MVCC columns while simultaneously locking them for the current transaction.
+      target_chunk->get_scoped_mvcc_data_lock()->grow_by(target_chunk_num_inserted_rows, context->transaction_id());
 
-      remaining_rows -= rows_to_insert_this_loop;
+      // Grow data Segments.
+      auto old_size = target_chunk->size();
+      for (ColumnID column_id{0}; column_id < target_chunk->column_count(); ++column_id) {
+        resolve_data_type(_target_table->column_data_type(column_id), [&](const auto data_type_t) {
```

Comment:
Use `BaseValueSegment`?

Reply:
To do what? If want to grow ValueSegment::values(), we need to know its type.

Reply:
Not if `BaseValueSegment` has a virtual `resize` method. Why should I have to know the type of a segment to resize it?

Reply:
No hard opinion on this, but there is only one place where we grow/resize segments: `Insert`. I'd rather not add a virtual function which is then only used in one place.I'd inline it.

Reply:
Works for me.